src/chain/chain-training.cc

```diff
@@ -54,9 +64,24 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
       numerator.Backward(xent_output_deriv);
     }
   }
+  
+  num_logprob_weighted += -opts.boost * supervision.num_sequences *
+                          supervision.frames_per_sequence;
+  CuMatrixBase<BaseFloat> *boosting_mask = nnet_output_deriv;
+  if (opts.boost != 0.0 && opts.hard_boost) {
+    boosting_mask = new CuMatrix<BaseFloat>(*nnet_output_deriv);
+    boosting_mask->ApplyHeaviside();
```

Comment:
looks fine, but it occurred to me that if the hard mask turns out to be better, we should probably (a) hardcode it to use the hard mask, and (b) we can give the soft version of the mask to the denominator computation and have it do the heaviside itself.  It creates a copy of the matrix anyway, to do the transpose, so we can have it copy the mask before the actual probs, do the heaviside, and then add in the actual probs.  but no need to do this before testing the idea out, the speed difference would be extremely tiny.